Introduce multi partition predicate [HZ-1347]

software-is-art opened the pull request:
This is a redux of [PR-18171](https://github.com/hazelcast/hazelcast/pull/18171). I'm continuing on from @ashley-taylor.

I've gone with a slightly different approach of making a backwards compatible change to the existing `PartitionPredicate` interface (instead of making an entirely new interface). This seemed lower risk as existing code with single partition keys will continue to work, while anything new can take advantage of the `getPartitionKeys` method.

Breaking changes (list specific methods/types/messages):
* PartitionPredicateImpl constructor

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [ ] Label `Add to Release Notes` or `Not Release Notes content` set
- [ ] Request reviewers if possible
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases
- [ ] New public APIs have `@Nonnull/@Nullable` annotations
- [ ] New public APIs have `@since` tags in Javadoc

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
[![CLA assistant check](https://cla.hazelcast.org/pull/badge/not_signed)](https://cla.hazelcast.org/hazelcast/hazelcast?pullRequest=21319) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.hazelcast.org/hazelcast/hazelcast?pullRequest=21319) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: ashley-taylor<br/>:x: software-is-art<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.hazelcast.org/check/hazelcast/hazelcast?pullRequest=21319) it.</sub>

software-is-art commented on hazelcast/src/main/java/com/hazelcast/query/impl/predicates/PartitionPredicateImpl.java:
```diff
@@ -49,23 +57,43 @@ public PartitionPredicateImpl() {
      */
-    public PartitionPredicateImpl(Object partitionKey, Predicate<K, V> target) {
-        this.partitionKey = checkNotNull(partitionKey, "partitionKey can't be null");
+    public PartitionPredicateImpl(Set<? extends Object> partitionKeys, Predicate<K, V> target) {
+        this.partitionKeys = checkNotNull(partitionKeys, "partitionKeys can't be null");
+        checkTrue(partitionKeys.size() > 0, "partitionKeys must not be empty");
```
I haven't included a backwards compatible constructor here, I can add one if this is a class that is directly interacted with by external consumers

ahmetmircik replied:
it is under impl package, there is no expected public usage.

vbekiaris commented:
run-lab-run

vbekiaris commented:
run-lts-compilers

software-is-art commented:
> The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/11033/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/11033/Hazelcast-pr-builder-11033.zip)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
> 
> Click to expand the log file

I'll need to back out changes made to the binary compatibility test file, and go from there

software-is-art commented:
@vbekiaris am I able to run the test suite using the same comments?

Holmistr commented:
@software-is-art Unfortunately no, due to security reasons (classic cliche :) ). Let me trigger it.

Holmistr commented:
run-lab-run

Holmistr commented:
Related issue: https://github.com/hazelcast/hazelcast/pull/21319

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java:
```diff
@@ -1391,14 +1395,19 @@ protected <T extends Result> T executeQueryInternal(Predicate predicate, Aggrega
+                Data key = toData(o);
+                int partitionId = partitionService.getPartitionId(key);
+                partitionIds.add(partitionId);
+                allFalse &= target.mode() == TargetMode.LOCAL_NODE && !partitionService.isPartitionOwner(partitionId)
```
Is it ok to create a short-circuit here? Seems after first `false`, loop can finish, instead of iterating whole given partition key set

software-is-art replied:
`allFalse` isn't used as the loop control variable.

This code is inside the `for (Object o : partitionPredicate.getPartitionKeys())` iteration, so it will be evaluated for every object in `getPartitionKeys`.

After the iterator has finished, this variable will be true if and only if all of the predicates would evaluate to false (either the target mode is local and the partition id isn't the owner, or the target mode is partition owner but the target doesn't contain the partition id)

mdumandag replied:
I didn't get why can't we do that. Were the comments for an old version of the implementation in this PR? I feel like we can short-circuit the loop with the current state

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java:
```diff
@@ -1391,14 +1395,19 @@ protected <T extends Result> T executeQueryInternal(Predicate predicate, Aggrega
+            PartitionIdSet partitionIds = new PartitionIdSet(partitionService.getPartitionCount());
+            boolean allFalse = true;
+            for (Object o : partitionPredicate.getPartitionKeys()) {
+                Data key = toData(o);
```
No need this conversion, `getPartitionId` already does it.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java:
```diff
@@ -687,14 +685,17 @@ protected void removeAllInternal(Predicate predicate) {
-                int partitionId = partitionService.getPartitionId(partitionKey);
+                PartitionIdSet partitionIds = new PartitionIdSet(partitionService.getPartitionCount());
+                partitionPredicate.getPartitionKeys().forEach(o -> {
+                    Data key = toData(o);
```
no need this, `getPartitionId` automatically does it.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/query/Predicates.java:
```diff
@@ -574,7 +577,21 @@ public static <K, V> PagingPredicate<K, V> pagingPredicate(Predicate<K, V> predi
+     * @param target       the target {@link Predicate}
+     * @param <K>          the type of keys the predicate operates on.
+     * @param <V>          the type of values the predicate operates on.
+     * @throws NullPointerException     if partition key or target predicate is {@code null}
```
please add a test for all possible validation exceptions thrown.

ahmetmircik commented:
@software-is-art can you please also add performance and unit tests from client perspective? 
Both client and server performance tests would be good to see no regression is introduced with this change, maybe perf tests also show us improvements.

software-is-art commented:
> @software-is-art can you please also add performance and unit tests from client perspective?
> Both client and server performance tests would be good to see no regression is introduced with this change, maybe perf tests also show us improvements.

@ahmetmircik is there an existing performance test you can point me to?

software-is-art commented:
@ahmetmircik I've added a [server side benchmark](hazelcast/src/test/java/com/hazelcast/map/PartitionsPredicatePerformanceTest.java) which I've run against this fork as well as the upstream master.

The maximum number of partitions was set to 16 for these tests. For upstream master, predicates are limited to hitting 1 or all 16 partitions.

The benchmark results show the difference in performance depends linearly on the number of partitions the predicate is invoked against and the predicate execution time.

The predicate execution time was varied using a spin wait. 

Each trial had a 10 second warmup followed by a 60 second run continually executing the aggregate with the given predicate.

From what I can tell, the results for the Single Partition case are within run-to-run variance between upstream and this PR.

### Upstream Master
Partitions Hit By Predicate|Items Per Partition|Predicate Load Factor (item seconds)|min (ms)|max (ms)|mean (ms)
| --- | --- | --- | --- | --- | --- |
1|1|1e-09|0.004083|4.882666|0.004890317734252377
1|1|1e-06|0.003333|2.334375|0.008971875385198128
1|1|0.001|1.0075|4.40175|1.0102355929707416
1|1|1.0|1000.037125|1000.06775|1000.0392459952085
16|1|1e-09|0.01125|3.520333|0.012036559755270515
16|1|1e-06|0.026792|2.728708|0.02753688780761281
16|1|0.001|16.012167|20.10125|16.0140898213244
16|1|1.0|16000.131458|16000.146291|16000.135369625

### This PR
Partitions Hit By Predicate|Items Per Partition|Predicate Load Factor (item seconds)|min (ms)|max (ms)|mean (ms)
| --- | --- | --- | --- | --- | --- |
1|1|1e-09|0.003916|8.405958|0.005547322495319188
1|1|1e-06|0.004958|14.337917|0.008564589096688044
1|1|0.001|1.007208|7.453791|1.0095187649462758
1|1|1.0|1000.037333|1000.060583|1000.0416434002087
2|1|1e-09|0.001917|1.823125|0.002196049152518402
2|1|1e-06|0.004208|10.441208|0.004482619341515382
2|1|0.001|2.002208|5.491208|2.0039896547091303
2|1|1.0|2000.02825|2000.03575|2000.0297931088842
4|1|1e-09|0.003541|3.028417|0.004028379505553561
4|1|1e-06|0.007458|5.317375|0.007858638501109353
4|1|0.001|4.003542|7.561417|4.003973636296994
4|1|1.0|4000.041167|4000.045833|4000.0424592127683
8|1|1e-09|0.007125|17.660125|0.007713850858933721
8|1|1e-06|0.014416|4.82775|0.015012679361262765
8|1|0.001|8.006583|11.780084|8.007771784074897
8|1|1.0|8000.065083|8000.073041|8000.0664706328125
16|1|1e-09|0.01225|5.606208|0.01320051131394354
16|1|1e-06|0.028042|6.503042|0.02937286093992011
16|1|0.001|16.0135|21.298292|16.01520113779588
16|1|1.0|16000.11675|16000.1285|16000.1202915

software-is-art commented:
@ahmetmircik I've run into issues running the simulator. Any chance you could run the simulations you're interested in? It looks like it will be some time until we get to the bottom of the issues in [slack](https://hazelcastcommunity.slack.com/archives/C015Q2TUBKL/p1653597883705329)

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java:
```diff
@@ -1288,13 +1286,13 @@ public void executeOnEntriesInternal(EntryProcessor entryProcessor, Predicate pr
                 PartitionPredicate partitionPredicate = (PartitionPredicate) predicate;
-                Data key = toData(partitionPredicate.getPartitionKey());
-                int partitionId = partitionService.getPartitionId(key);
+                PartitionIdSet partitionIds = new PartitionIdSet(partitionService.getPartitionCount());
```
Can you please also do this change for client protocol related classes? For instance `MapExecuteWithPredicateMessageTask`, now this class is executed on a single partition independent of the provided partition set size?

software-is-art replied:
I think that's the only class that needs changing. The rest of the client code uses the partition id to determine which cluster member to invoke the request on. `getPartitionKey` selects a random key from the set of partition keys on `PartitionPredicate` so it should have the desired result of load balancing.

ahmetmircik commented:
@software-is-art can you please rebase and address the issue in my latest comment?

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientMapProxy.java:
```diff
@@ -165,24 +180,8 @@
-import static java.util.Collections.emptyMap;
-import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
```
better to rollback the changes in this file to make PR clearer, as far as i see, only import order was changed here.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/client/impl/spi/ClientProxy.java:
```diff
@@ -26,12 +28,9 @@
-import static com.hazelcast.internal.util.ExceptionUtil.rethrow;
+import javax.annotation.Nonnull;
 
 /**
```
same import issue.

ahmetmircik commented:
run-lab-run

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/11615/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/11615/Hazelcast-pr-builder-11615.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.2:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 2 errors reported by Checkstyle 8.38 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/internal/partition/impl/InternalPartitionServiceLiteMemberTest.java:40:1: Duplicate import to line 31 - java.util.List. [RedundantImport]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/internal/partition/impl/InternalPartitionServiceLiteMemberTest.java:41:1: Duplicate import to line 32 - java.util.Map. [RedundantImport]
--------------------------
</pre>
</details>

ahmetmircik commented:
@software-is-art there are still 2 issues: 
1. checkstyle errors, please run `mvn -e --no-transfer-progress -B clean test-compile` and fix the problems shown.
2. `git rebase master` complaining from conflicts.

Can you please fix these ones?

ahmetmircik commented:
run-lab-run

ahmetmircik commented:
@software-is-art While testing i have noticed `AbstractMapQueryMessageTask` has not been fixed, it still works on a single partition. Please check all places where `instanceof PartitionPredicate` is used and do required changes.

Also `git rebase master` is still showing conflicts, maybe you can create a fresh branch from current master and apply your changes on top of it to make things clearer.

software-is-art commented:
@ahmetmircik I could only find one more place to change in `AbstractMapQueryMessageTask`

I can't see any merge conflicts when I rebase locally

ahmetmircik commented:
Thanks for your work, seems you have covered all issues that i can see. LGTM.

ahmetmircik commented on extensions/cdc-mysql/src/test/java/com/hazelcast/jet/cdc/mysql/MySqlCdcWhiteBlackListIntegrationTest.java:
```diff
@@ -347,13 +347,13 @@ private List<String> allTable0ExpectedOperations(int dbSuffix) {
         int id3 = 1000 * dbSuffix + 3;
         int id4 = 1000 * dbSuffix + 4;
         return Arrays.asList(
-                id1 + "/0:INSERT:TableRow {id=" + id1 + ", value1=" + db + "_table0_val1_0, "
```
why do we need the changes in this file? Please rollback, if they are leftover.

software-is-art replied:
Force pushed a rebase which should fix this

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/jet/impl/JobRepository.java:
```diff
@@ -537,6 +538,18 @@ Set<Long> getAllJobIds() {
         return ids;
     }
 
+    public Collection<String> getActiveJobNames() {
```
leftover?

ahmetmircik commented on hazelcast/src/test/java/example/serialization/TestDeserialized.java:
```diff
@@ -26,6 +26,7 @@ public class TestDeserialized implements Serializable {
     public static volatile boolean isDeserialized = false;
 
     private void writeObject(java.io.ObjectOutputStream out) throws IOException {
+
```
leftover?

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/query/Predicates.java:
```diff
@@ -571,10 +574,25 @@ public static <K, V> PagingPredicate<K, V> pagingPredicate(Predicate<K, V> predi
+     * @throws NullPointerException     if partitionKeys or target predicate are {@code null}
+     * @throws IllegalArgumentException if partitionkeys is an empty set
+     */
+    public static <K, V> PartitionPredicate<K, V> partitionsPredicate(Set<? extends Object> partitionKeys,
```
What about changing name of this method to `multiPartitionPredicate`? Goal is having less error prone naming and making it easier to read for eyes.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/query/impl/predicates/PartitionPredicateImpl.java:
```diff
@@ -95,22 +123,32 @@ public int getClassId() {
     public void writeData(ObjectDataOutput out) throws IOException {
-        out.writeObject(partitionKey);
+        if (out.getVersion().isLessThan(Versions.V5_2)) {
+            out.writeObject(partitionKey);
```
if version is less than 5.2, we are sending null predicate in this PR, this should be fixed by sending the single key in `partitionKeys`. Otherwise existing code will be broken if one upgrades 5.2. Even your hazelcast is 5.2, your cluster can be 5.1 so we must keep compatibility.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/query/impl/predicates/PartitionPredicateImpl.java:
```diff
@@ -35,11 +41,13 @@
     private static final long serialVersionUID = 1L;
+    private static final Random RANDOM = new Random();
 
     private Object partitionKey;
```
can we delete `partitionKey` and rely on `partitionKeys` in all relevant places?

software-is-art replied:
Yes that sounds good to me. I wasn't sure if we wanted a fast path or not.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/query/impl/predicates/PartitionPredicateImpl.java:
```diff
@@ -124,15 +162,15 @@ public boolean equals(Object o) {
         PartitionPredicateImpl<?, ?> that = (PartitionPredicateImpl<?, ?>) o;
 
-        if (partitionKey != null ? !partitionKey.equals(that.partitionKey) : that.partitionKey != null) {
+        if (partitionKeys != null ? !partitionKeys.equals(that.partitionKeys) : that.partitionKeys != null) {
```
Can be simplified like this: ` if (!Objects.equals(partitionKeys, that.partitionKeys)) {`?

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/query/impl/predicates/PartitionPredicateImpl.java:
```diff
@@ -124,15 +162,15 @@ public boolean equals(Object o) {
+        if (partitionKeys != null ? !partitionKeys.equals(that.partitionKeys) : that.partitionKeys != null) {
             return false;
         }
         return target != null ? target.equals(that.target) : that.target == null;
```
same simplification: `Objects.equals(target, that.target);`

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/query/impl/predicates/PredicateDataSerializerHook.java:
```diff
@@ -152,6 +152,11 @@ public IdentifiedDataSerializable createNew(Integer arg) {
                 return new PartitionPredicateImpl();
             }
         };
+        constructors[PARTITIONS_PREDICATE] = new ConstructorFunction<Integer, IdentifiedDataSerializable>() {
```
we don't need this change, previous `PARTITION_PREDICATE ` should be sufficient.

software-is-art replied:
Good catch. A leftover from the prior implementation attempt.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractInvocationFuture.java:
```diff
@@ -1250,9 +1250,10 @@ protected void onComplete() {
     private void warnIfSuspiciousDoubleCompletion(Object s0, Object s1) {
         if (s0 != s1 && !(isStateCancelled(s0)) && !(isStateCancelled(s1))) {
-            logger.warning(String.format("Future.complete(Object) on completed future. "
+            String message = String.format("Future.complete(Object) on completed future. "
```
irrelevant change, can you please rollback it?

ahmetmircik reviewed (changes requested):
Thanks for your work, seems you have covered all issues that i can see in perf test. 
LGTM from this perspective.
But i can still see some PR specific issues, please also address them.

software-is-art commented:
@ahmetmircik I think I've managed to clean up the PR correctly this time.

My IDE was auto merging my origin branch which was undoing the rebase on push. Force pushed this time instead of letting IntelliJ handle it.

Let me know if I've missed anything else.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/query/PartitionPredicate.java:
```diff
@@ -33,10 +35,26 @@
+     *
+     * @return the partition keys
+     */
+    default Collection<? extends Object> getPartitionKeys() {
```
Seems it is not required to have a default implementation for `getPartitionKeys`. Since in `PartitionPredicateImpl` we always override it.

Having only this should be enough:
```
    Collection<? extends Object> getPartitionKeys();
```

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/query/impl/predicates/PartitionPredicateImpl.java:
```diff
@@ -121,19 +156,24 @@ public boolean equals(Object o) {
+        throw new RuntimeException(
+            "Unreachable branch. PartitionPredicateImpl constructor should check and throw if partitionKeys is empty"
+        );
+    }
```
Maybe we can add a javaDoc for method's purpose and an assertion to catch unexpected cases.
```
    /**
     * This method is here only for backward compatibility
     * with versions < 5.2.
     * @return a single partition key.
     */
    private Object getSinglePartitionKey() {
        assert partitionKeys.size() == 1;
        
for (Object key : partitionKeys) {
            return key;
        }
        throw new RuntimeException(
            "Unreachable branch. PartitionPredicateImpl constructor should check and throw if partitionKeys is empty"
        );
    }
```

software-is-art replied:
JavaDoc for a `private` method seems a bit overkill

ahmetmircik commented:
run-lab-run

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/11739/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/11739/Hazelcast-pr-builder-11739.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   ClientPartitionPredicateTest.keys_withPagingPredicate:111 ? ClassCast java.uti...
[ERROR]   ClientPartitionPredicateTest.values_withPagingPredicate:97 ? ClassCast java.ut...
[INFO] 
[ERROR] Tests run: 50588, Failures: 0, Errors: 2, Skipped: 238
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

ahmetmircik commented:
@software-is-art 

We also have 2 failing tests: 
- `ClientPartitionPredicateTest.keys_withPagingPredicate`
- `ClientPartitionPredicateTest.values_withPagingPredicate`

software-is-art commented on hazelcast/src/main/java/com/hazelcast/query/impl/predicates/PartitionPredicateImpl.java:
```diff
@@ -49,23 +56,43 @@ public PartitionPredicateImpl() {
+     * @return the randomly selected partition key from {@linkplain #getPartitionKeys()} partitionKeys}.
      */
     @Override
     public Object getPartitionKey() {
```
The implementation is internal, but the interface is public. I'm concerned that if we throw the exception we may end up surprising consumers of the interface.

If the exception were thrown only if we have a set size > 1 that might be ok, as that will still allow the API to be backwards compatible in the single key case.

ahmetmircik replied:
I think you are right for backward compatibility concern, only throwing an exception for the case where a set size > 1 should be sufficient.

mdumandag commented on hazelcast/src/main/java/com/hazelcast/query/Predicates.java:
```diff
@@ -571,10 +574,25 @@ public static <K, V> PagingPredicate<K, V> pagingPredicate(Predicate<K, V> predi
+     * Creates a new partition predicate that restricts the execution of the target predicate to a subset of partitions.
+     *
+     * @param partitionKeys the partition keys
+     * @param target       the target {@link Predicate}
```
minor: maybe add one space for the each param after the partitionKeys so that the docs would be in-line

mdumandag commented on hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/holder/PagingPredicateHolder.java:
```diff
@@ -116,7 +116,7 @@ private static <K, V> PagingPredicateHolder ofInternal(@Nonnull PartitionPredica
         PagingPredicateImpl<K, V> pagingPredicate = (PagingPredicateImpl<K, V>) partitionPredicate.getTarget();
 
-        Data partitionKeyData = serializationService.toData(partitionPredicate.getPartitionKey());
+        Data partitionKeyData = serializationService.toData(partitionPredicate.getPartitionKeys());
```
I feel like to make it work, we need to add a new parameter to the https://github.com/hazelcast/hazelcast-client-protocol/blob/master/protocol-definitions/custom/Custom.yaml#L852 with the type of `List_Data`, and set both the `partitionKeyData` and the new parameter so that both the old and new members can continue working. Note that, the `since` value of the parameter should be `2.5` for the 5.2 release

mdumandag replied:
A similar thing applies to the read path (where we read client messages from the old clients like 5.0)

software-is-art replied:
That's exactly what has caused the two tests to fail, and why I was originally so cautious about touching the client code (it would suffice to have server side support only for our usecase).

Do you think it's worth doing the client side code in a separate PR @mdumandag?

mdumandag replied:
I would vote for doing it here. WDYT @ahmetmircik?

ahmetmircik replied:
we can do it here, no need a separate PR.

software-is-art replied:
I've managed to get the codecs working with backwards compatible request/response.

The bug causing these two unit test failures isn't because the return type is incompatible (MapKeySet already returns List_Data because querying a single partition can still result in multiple return values).

The bug was introduced when I changed `AbstractMapQueryMessageTask.invokeOnPartitions` to use the map service query engine for a keyset:

```QueryResult result = mapServiceContext.getQueryEngine(name).execute(query, Target.createPartitionTarget(query.getPartitionIdSet()));```

When this code is executed it adjusts the predicate specified on the query in `QueryEngineImpl.adjustQuery` 
```private Query adjustQuery(Query query) {
        IterationType retrievalIterationType = getRetrievalIterationType(query.getPredicate(), query.getIterationType());
        Query adjustedQuery = Query.of(query).iterationType(retrievalIterationType).build();
        if (adjustedQuery.getPredicate() instanceof PagingPredicateImpl) {
            ((PagingPredicateImpl) adjustedQuery.getPredicate()).setIterationType(query.getIterationType());
        } else {
            if (adjustedQuery.getPredicate() == Predicates.alwaysTrue()) {
                queryResultSizeLimiter.precheckMaxResultLimitOnLocalPartitions(adjustedQuery.getMapName());
            }
        }
        return adjustedQuery;
    }```

This adjustment changes `iterationType` on `PagingPredicateImpl` instance to match the iteration type specified by the query object (in this case it changes KEY to ENTRY). 

Looking at the code `MapKeySetWithPagingPredicateTask.getIterationType` is implemented as: 
```@Override
    protected IterationType getIterationType() {
        return IterationType.ENTRY;
    }``` 

This means there has always been an inconsistency between the `PagingPredicateImpl` and `MapKeySetWithPagingPredicateTask`, because the old code never executed a map query via the service it was never noticed.

I'm proposing we change ```MapKeySetWithPagingPredicateTask.getIterationType``` to return `IterationType.KEY`

mdumandag commented on hazelcast/src/main/java/com/hazelcast/query/PartitionPredicate.java:
```diff
@@ -33,10 +35,26 @@
+     * that wraps the {@linkplain #getPartitionKey() partition key} in a singleton
+     * collection is provided for backwards compatibility.
+     *
+     * @return the partition keys
```
minor: we can add `@since` here for this method

mdumandag commented on hazelcast/src/main/java/com/hazelcast/query/impl/predicates/PartitionPredicateImpl.java:
```diff
@@ -95,22 +122,30 @@ public int getClassId() {
     @Override
     public void readData(ObjectDataInput in) throws IOException {
-        this.partitionKey = in.readObject();
+        if (in.getVersion().isLessThan(Versions.V5_2)) {
```
I don't think this is also going to work. 

From the paths I see, the versions are only set to identified header in the enterprise edition and are simply skipped in the open-source implementation while reading. See https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/DataSerializableSerializer.java#L155-L158 

So, the version you will see here will always be the version of the _current codebase_, not the version that sent this data, if you are using open-source. Do you guys do anything special for such cases @ahmetmircik?

mdumandag replied:
One way to solve this without version checks would be introducing a new implementation(MultiPartitionPredicateImpl) and keeping this one, and giving it a unique class id so that the old readers cannot read it. And that would be fine I suppose, as we don't expect old servers to work with `multiPartitionPredicate` because it is introduced in 5.2

software-is-art replied:
Two implementations of the interface sounds ok to me

software-is-art replied:
How does this work in other operations that use this pattern already? e.g. [PutAllOperation](https://github.com/hazelcast/hazelcast/blob/14ef0dbefa4a6fa00a50a8873aeea02ec6c3f5e4/hazelcast/src/main/java/com/hazelcast/map/impl/operation/PutAllOperation.java#L216-L234) uses `getVersion()` to optionally read/write a boolean

mdumandag replied:
The other places are fine. They are part of an enterprise feature called Rolling Upgrade which lets the adjacent minor versions of the Hazelcast members form a cluster during upgrades. 

But, this class is re-used by the client and members. Clients serialize this class and send it as part of the requests. Imagine that you are an open-source Hazelcast user, you have 5.2 servers and 5.0 clients. Now, these checks won't work, and you will try to read a single partition key as a set of partition keys. Similar thing holds true for 5.2 clients and 5.0 servers.

So, in short, if the class can be serialized by both members and clients, we should not rely on these version checks.

software-is-art replied:
@mdumandag I've created a `MultiPartitionPredicate` class. I haven't pushed it yet, as it's taking a while to run through all the tests and make sure I've not missed something. 

Will hopefully have this change pushed tomorrow.

software-is-art replied:
I've pushed what I have in the meantime if you want to place some early feedback @mdumandag / @ahmetmircik

mdumandag commented on hazelcast/src/main/java/com/hazelcast/query/impl/predicates/PartitionPredicateImpl.java:
```diff
@@ -95,22 +122,30 @@ public int getClassId() {
     @Override
     public void readData(ObjectDataInput in) throws IOException {
-        this.partitionKey = in.readObject();
+        if (in.getVersion().isLessThan(Versions.V5_2)) {
```
I don't think this is also going to work. 

From the paths I see, the versions are only set to identified header in the enterprise edition and are simply skipped in the open-source implementation while reading. See https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/DataSerializableSerializer.java#L155-L158 

So, the version you will see here will always be the version of the _current codebase_, not the version that sent this data, if you are using open-source. Do you guys do anything special for such cases @ahmetmircik?

pveentjer commented:
I don't see the added value of this. If you want to execute on multiple partitions, then call the method multiple times with different PartitionedPredicates. No need to make the API more complex for an edge case that can easily be handled differently.

software-is-art commented:
> I don't see the added value of this. If you want to execute on multiple partitions, then call the method multiple times with different PartitionedPredicates. No need to make the API and implementation more complex for an edge case that can easily be handled differently.
> 
> PartitionPredicate is legacy anyway.

@pveentjer respectfully, we have to disagree.

Our usecase at Phocas is we're storing large data per key in the map and performing aggregations on these.

The predicate takes much longer to execute for us as a result. If we know we only need to hit a subset of partitions its a massive time saving for us.

Separate queries are not enough to achieve this. If we have to hit every node with a separate query this aggregation cannot happen locally to each node. Rather, everything will be serialised to the master node which is far too expensive to be practical given the size of the objects.

Ultimately this is a latency sensitive application, and the extra time/memory burden of invoking multiple queries and serializing is not acceptable. We already have a monkey patched version of Hazelcast in our application that demonstrates the value to us, this is ultimately about upstreaming that so others can benefit and enabling us to upgrade to `5.2` when it comes out.

I don't think the API will become more complex as specifying a set is more natural than specifying a single value. If we're careful it should easily be backwards compatible.

ahmetmircik commented:
run-lab-run

software-is-art commented:
@ahmetmircik the codec had been modified by my IDE. I disabled optimise imports and regenerated the file. Test now passes

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/11898/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/11898/Hazelcast-pr-builder-11898.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   ClientMapBasicTest.github_11489_verifyNoFailingCastOnEntriesForPagingPredicate:1360->github_11489_verifyNoFailingCastOnValue:1516->lambda$github_11489_verifyNoFailingCastOnEntriesForPagingPredicate$14:1360 ? NullPointer
[ERROR]   ClientMapBasicTest.github_11489_verifyNoFailingCastOnKeySetForPagingPredicate:1349->github_11489_verifyNoFailingCastOnValue:1516->lambda$github_11489_verifyNoFailingCastOnKeySetForPagingPredicate$12:1349 ? NullPointer
[ERROR]   ClientMapBasicTest.github_11489_verifyNoFailingCastOnValueForPagingPredicate:1338->github_11489_verifyNoFailingCastOnValue:1516->lambda$github_11489_verifyNoFailingCastOnValueForPagingPredicate$10:1338 ? NullPointer
[ERROR]   ClientMapBasicTest.testMapClonedCollectionsImmutable:1014 ? NullPointer
[ERROR]   ClientMapIssueTest.testMapPagingEntries:194 ? NullPointer
[ERROR]   ClientMapIssueTest.testMapPagingKeySet:316 ? NullPointer
[ERROR]   ClientMapIssueTest.testMapPagingValues:218 ? NullPointer
[ERROR]   ClientPagingPredicateTest>PagingPredicateTest.betweenPagingPredicateWithEmployeeTest:375->PagingPredicateTest.pagingPredicateWithEmployeeObjectTest:390 ? NullPointer
[ERROR]   ClientPagingPredicateTest>PagingPredicateTest.equalsPredicateWithEmployeeTest:473->PagingPredicateTest.pagingPredicateWithEmployeeObjectTest:390 ? NullPointer
[ERROR]   ClientPagingPredicateTest>PagingPredicateTest.lessThanPredicateWithEmployeeTest:450->PagingPredicateTest.pagingPredicateWithEmployeeObjectTest:390 ? NullPointer
[ERROR]   ClientPagingPredicateTest>PagingPredicateTest.mapPagingPredicateEmployeeObjectWithOrderedIndexLargeTest:411->PagingPredicateTest.mapPagingPredicateEmployeeObjectWithOrderedIndex:424 ? NullPointer
[ERROR]   ClientPagingPredicateTest>PagingPredicateTest.mapPagingPredicateEmployeeObjectWithOrderedIndexSmallTest:406->PagingPredicateTest.mapPagingPredicateEmployeeObjectWithOrderedIndex:424 ? NullPointer
[ERROR]   ClientPagingPredicateTest>PagingPredicateTest.testCustomComparatorAbleToActOnKeysAndValues:335 ? NullPointer
[ERROR]   ClientPagingPredicateTest>PagingPredicateTest.testEmptyIndexResultIsNotCausingFullScan:323 ? NullPointer
[ERROR]   ClientPagingPredicateTest>PagingPredicateTest.testEqualValuesPaging:241 ? NullPointer
[ERROR]   ClientPagingPredicateTest>PagingPredicateTest.testGoToPreviousPageBeforeTheStart:262 ? NullPointer
[ERROR]   ClientPagingPredicateTest>PagingPredicateTest.testKeyPaging:215 ? NullPointer
[ERROR]   ClientPagingPredicateTest>PagingPredicateTest.testLargePageSizeIsNotCausingIndexOutBoundsExceptions:308 ? NullPointer
[ERROR]   ClientPagingPredicateTest>PagingPredicateTest.testNextPageAfterResultSetEmpty:275 ? NullPointer
[ERROR]   ClientPagingPredicateTest>PagingPredicateTest.testPagingWithFilteringAndComparator:172 ? NullPointer
[ERROR]   ClientPagingPredicateTest>PagingPredicateTest.testPagingWithFilteringAndComparatorAndIndex:191 ? NullPointer
[ERROR]   ClientPagingPredicateTest>PagingPredicateTest.testPagingWithoutFilteringAndComparator:154 ? NullPointer
[ERROR]   ClientPagingPredicateTest>PagingPredicateTest.testWithoutAnchor:137 ? NullPointer
[ERROR]   ClientPartitionPredicateTest.values_withPagingPredicate:97 ? ClassCast java.ut...
[INFO] 
[ERROR] Tests run: 50588, Failures: 0, Errors: 24, Skipped: 238
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

software-is-art commented:
@ahmetmircik with some further research it looks like the map task iteration types should remain as ENTRY.

That leaves the only viable option to fix the incorrect iteration type on the predicate as cloning the predicate before adjusting it in `QueryEngineImpl`.

I've added a `clone()` method to `PartitionPredicateImpl` which protects the caller's predicate from getting its iteration type clobbered.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/holder/PagingPredicateHolder.java:
```diff
@@ -39,17 +43,24 @@ public class PagingPredicateHolder {
+    public PagingPredicateHolder(AnchorDataListHolder anchorDataListHolder, Data predicateData, Data comparatorData,
+                                 int pageSize, int page, byte iterationTypeId,
+                                 Data partitionKeyData,
+                                 boolean partitionKeysDataExists,
```
`partitionKeysDataExists` is not used?

software-is-art replied:
No it's not, but it has to be there as this is called by auto generated code

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryEngineImpl.java:
```diff
@@ -125,15 +125,16 @@ public Result execute(Query query, Target target) {
+        Query.QueryBuilder builder = Query.of(query).iterationType(retrievalIterationType);
+        if (query.getPredicate() instanceof PagingPredicateImpl) {
+            // PagingPredicateImpl must be cloned before adjusting else it corrupts the iteration type for the caller
+            PagingPredicateImpl clonedPredicate = new PagingPredicateImpl((PagingPredicateImpl) query.getPredicate());
```
Can you please remind what specific test is failing if you don't clone?

software-is-art replied:
[These ones](https://github.com/hazelcast/hazelcast/pull/21319#issuecomment-1168543103)

The problem is the predicate iteration type is modified by downstream code so we must clone it to make sure it doesn't corrupt the caller.

ahmetmircik replied:
@software-is-art i see the failing tests thanks. But why do this new multi-partition-predicate related changes require this cloning? Previously, there was no need. Could you please also tell this?

software-is-art replied:
It's not that previously there was no need. This bug has always been in this code from what I can tell.

What prevented it from occurring previously was that we didn't use the QueryEngine at all in the paging predicate path for the special case where it contained a `PartitionPredicate`.

mdumandag commented:
@ahmetmircik @software-is-art I didn't take a look at the recent changes after my initial review, but I see that this comment is not addressed yet. https://github.com/hazelcast/hazelcast/pull/21319#discussion_r916586999. I think solving this is a must for backward compatibility

ahmetmircik commented:
Hi @software-is-art, to have this in 5.2, we need to go a bit faster, could you please address latest reviews?

vbekiaris commented:
@software-is-art can you please address reviewers' comments and resolve conflicts? Feature freeze for 5.2-BETA is approaching, we should get this PR done soon in order to make it for 5.2.

software-is-art commented on hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalPartitionServiceImpl.java:
```diff
@@ -872,6 +873,13 @@ public IPartition[] getPartitions() {
     }
 
+    @Override
+    public PartitionIdSet getPartitionIdSet(Collection<Object> keys) {
```
It does only use members of `IPartitionSerivce`. I've added a `default` implementation on the interface instead of implementing it here.

ahmetmircik commented:
run-lab-run

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12306/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12306/Hazelcast-pr-builder-12306.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   QueueDestroyTest.checkStatsMapEntryRemovedWhenQueueDestroyed:78 expected:<0> but was:<1>
[ERROR]   PartitionPredicateTest.testToString:199->HazelcastTestSupport.assertContains:866 'PartitionPredicate{partitionKey=8ddc8882-4c86-44ea-be80-732b8d18e1f3, target=TruePredicate{}}' didn't contain expected 'partitionKeys=[8ddc8882-4c86-44ea-be80-732b8d18e1f3]'
[INFO] 
[ERROR] Tests run: 50613, Failures: 2, Errors: 0, Skipped: 238
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

mdumandag commented on hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java:
```diff
@@ -687,14 +685,16 @@ protected void removeAllInternal(Predicate predicate) {
+                operationService.invokeOnPartitions(
+                        SERVICE_NAME,
+                        operation,
+                        partitionService.getPartitionIdSet(partitionPredicate.getPartitionKeys())
```
In the old implementation, we were using the per map config partitioning strategy with `toDataWithStrategy`, but now the implementation uses regular `toData` with a global partitioning strategy. I feel like this can be a bit problematic. WDYT @ahmetmircik ?

software-is-art replied:
`partitionKey` was only ever used to get the `partitionId` to execute on.

From what I can see the data object isn't used for anything else.

I might be missing something, but because we now have the `getPartitionKeys()` method available this is no longer needed.

ahmetmircik replied:
`toDataWithStrategy` is used to find correct partition-ids. 
The returned object keys from `getPartitionKeys` method must be converted to data by using `toDataWithStrategy`, then partition-id-set must be populated by using these data.

mdumandag commented on hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java:
```diff
@@ -1288,13 +1288,15 @@ public void executeOnEntriesInternal(EntryProcessor entryProcessor, Predicate pr
             Map<Integer, Object> results;
             if (predicate instanceof PartitionPredicate) {
                 PartitionPredicate partitionPredicate = (PartitionPredicate) predicate;
-                Data key = toData(partitionPredicate.getPartitionKey());
```
Regarding my comment above, I don't understand why we were not using `toDataWithStrategy` to serialize the partition key in the old implementation here and in some other places. Maybe a bug in our end or using the `toDataWithStrategy` above `removeAllInternal` method was a mistake? I feel like the former. WDYT @ahmetmircik

ahmetmircik replied:
yeah, this seems a bug.

mdumandag commented on hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryEngineImpl.java:
```diff
@@ -125,15 +125,16 @@ public Result execute(Query query, Target target) {
-                queryResultSizeLimiter.precheckMaxResultLimitOnLocalPartitions(adjustedQuery.getMapName());
-            }
+        Query.QueryBuilder builder = Query.of(query).iterationType(retrievalIterationType);
+        if (query.getPredicate() instanceof PagingPredicateImpl) {
```
What about `MultiPartitionPredicateImpl`?

software-is-art replied:
I think in this case we will be ok as this is `PagingPredicateImpl` we are checking for.

From what I can tell, this wraps the inner predicate so when we clone this we only clone the `PagingPredicateImpl` to prevent the iteration type from being clobbered. We don't clone the inner predicate which would require the check for `Multi/PartitionPredicateImpl`

mdumandag commented on hazelcast/src/main/java/com/hazelcast/query/impl/predicates/MultiPartitionPredicateImpl.java:
```diff
@@ -0,0 +1,169 @@
+ * @param <V> type of the entry value
+ */
+@BinaryInterface
+public class MultiPartitionPredicateImpl<K, V> implements PartitionPredicate<K, V>, IdentifiedDataSerializable, Versioned {
```
This shouldn't be `Versioned` anymore

ahmetmircik replied:
A potential issue here: 
What happens when i pass `MultiPartitionPredicateImpl`, for example, to `executeOnEntries` method in a cluster with version 5.1? I expect it not to work.

mdumandag replied:
Why should it work? I believe it is okay for it to not work in this case, as the feature is implemented in 5.2. Similarly, if you send this predicate to 5.0 member from 5.2 client, I believe it shouldn't work as well (which is also the case with the current implementation)

ahmetmircik replied:
I expect it not to work, my above comment is saying that.

mdumandag replied:
Yeah, I see. I am just saying that I don't see this as a potential issue

software-is-art replied:
Good point, I will remove this.

ahmetmircik replied:
@software-is-art What about this one?
> What happens when i pass MultiPartitionPredicateImpl, for example, to executeOnEntries method in a cluster with version 5.1?

Do we throw feature is not supported exception in this case?

software-is-art replied:
Is there a standard way of doing that? Would I just be checking for the destination cluster version before sending it?

ahmetmircik replied:
You can check `MultiMapProxyImpl#putAllAsync` method to see how this control can be done.
Basically you need to do this control for source cluster entry points where one can pass multi-partition predicate, including client-message tasks.

I think, to address this you can create a separate PR later, after this one is merged, since this one is already got bloated.

software-is-art replied:
That sounds good to me, happy to have that in a separate PR

mdumandag commented on hazelcast/src/main/java/com/hazelcast/query/impl/predicates/MultiPartitionPredicateImpl.java:
```diff
@@ -0,0 +1,169 @@
+@BinaryInterface
+public class MultiPartitionPredicateImpl<K, V> implements PartitionPredicate<K, V>, IdentifiedDataSerializable, Versioned {
+
+    private static final long serialVersionUID = 1L;
```
Why do we need this?

software-is-art replied:
Copy/pasta. Will remove.

mdumandag commented on hazelcast/src/test/java/com/hazelcast/map/PartitionPredicateTest.java:
```diff
@@ -206,10 +205,20 @@ public void testSerialization() {
+    public void partitionPredicateWithNullPredicateThrowsNullPointerException() {
+        assertThrows(NullPointerException.class, () -> Predicates.partitionPredicate("foo", null));
+    }
+
```
should we add the same tests for multi partition predicate as well?

software-is-art replied:
I already have test for these in `PartitionsPredicateTest.java`

I'm renaming the class to `MultiPartitionPredicateTest.java` so look for `multiPartitionPredicateWithNullSetThrowsNullPointerException` on my next push

mdumandag commented on hazelcast/src/test/java/com/hazelcast/map/PartitionsPredicatePerformanceTest.java:
```diff
@@ -0,0 +1,184 @@
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({NightlyTest.class})
+public class PartitionsPredicatePerformanceTest extends HazelcastTestSupport {
```
I am not sure how running this test in nightly will be useful. I don't think anyone will take a look at the logs produced by this test and compare it to some baseline etc.

software-is-art replied:
Ok, are you suggesting I remove it or change the category?

mdumandag replied:
I believe the benchmark you have written is definitely useful, so I wouldn't recommend removing it altogether. I would just remove the test related annotations
```
@RunWith(HazelcastParallelClassRunner.class)
@Category({NightlyTest.class})
```
so that the test runner wouldn't pick it, Maybe even add a main method to it, instead of this test structure. I see that it is done for a few other files like `MapWordCountAggregationBenchmark` or `ClientQueuePerformanceBenchmark`

mdumandag replied:
Minor: Maybe we can do this as well

mdumandag commented on hazelcast/src/test/java/com/hazelcast/map/PartitionsPredicateTest.java:
```diff
@@ -0,0 +1,257 @@
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class PartitionsPredicateTest extends HazelcastTestSupport {
```
Can we have the same test for client as well? Maybe extend this class and provide necessary abstractions to use either member or client as a driver.

software-is-art replied:
Done

software-is-art commented on hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/holder/PagingPredicateHolder.java:
```diff
@@ -136,8 +163,14 @@ private static <K, V> PagingPredicateHolder buildHolder(SerializationService ser
         return new PagingPredicateHolder(anchorDataListHolder, predicateData, comparatorData, pagingPredicate.getPageSize(),
-                pagingPredicate.getPage(), pagingPredicate.getIterationType().getId(), partitionKeyData);
+                pagingPredicate.getPage(), pagingPredicate.getIterationType().getId(), null,
+                partitionKeysDataExists,
```
Ah gotcha, it's indicating that the frame is present not the data it contains. I thought it seemed redundant the way I understood it. 

Will change the logic to depend only on the `null` value.

software-is-art commented on hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/holder/PagingPredicateHolder.java:
```diff
@@ -77,7 +89,15 @@ public byte getIterationTypeId() {
+        if (partitionKeysData == null) {
+            return null;
+        }
+        // PartitionPredicateImpl ensures that partitionKeysData is non-empty
```
Updated to reference the `PagingPredicateHolder` constructor which provides this guarantee

software-is-art commented:
A few tests failing in the new client tests for multi partition predicate. Haven't figured out yet why they don't work for the client, but do for a local instance.

ahmetmircik commented:
run-lab-run

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12360/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12360/Hazelcast-pr-builder-12360.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   MultiPartitionPredicateClientTest>MultiPartitionPredicateTestSupport.removeAll:193 expected:<160> but was:<180>
[ERROR] Errors: 
[ERROR]   MultiPartitionPredicateClientTest>MultiPartitionPredicateTestSupport.testSerialization:236 ? IllegalArgument
[INFO] 
[ERROR] Tests run: 50626, Failures: 1, Errors: 1, Skipped: 238
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

software-is-art commented:
> 

I've determined that the `removeAll` seems to only be removing half of the expected items compared to the local instance. Digging into this today.

software-is-art commented:
I've fixed the remaining tests.

`MapRemoveAllMessageTask` was submitting the remove operation to only a single partition.

software-is-art commented on hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapRemoveAllMessageTask.java:
```diff
@@ -52,36 +49,19 @@ protected void processMessage() {
         OperationFactory operationFactory = createOperationFactory();
         OperationServiceImpl operationService = nodeEngine.getOperationService();
 
-        // We are running on a partition thread now and we are not allowed
-        // to call invokeOnPartitions(Async) on operation service because of
-        // that.
```
This is the only part I'm hesitant about, this comment indicates what I've done may not be desired.

Is there a better way to invoke on  partitions from here?

ahmetmircik replied:
I think it is a left-over comment, this part is not running on partition thread.
@mdumandag can you please confirm if this is the case?

mdumandag replied:
Yeah, this message task (and hence the `processMessage` method) is executed in the `hz:client` executor.

software-is-art commented:
@ahmetmircik could you please run the PR builder?

vbekiaris commented:
run-lab-run

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12427/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12427/Hazelcast-pr-builder-12427.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   BinaryCompatibilityTest.readAndVerifyBinaries:127
[ERROR]   BinaryCompatibilityTest.readAndVerifyBinaries:127
[ERROR]   BinaryCompatibilityTest.readAndVerifyBinaries:127
[ERROR]   BinaryCompatibilityTest.readAndVerifyBinaries:127
[ERROR]   BinaryCompatibilityTest.readAndVerifyBinaries:127
[ERROR]   BinaryCompatibilityTest.readAndVerifyBinaries:127
[ERROR]   BinaryCompatibilityTest.readAndVerifyBinaries:127
[ERROR]   BinaryCompatibilityTest.readAndVerifyBinaries:127
[INFO] 
[ERROR] Tests run: 4313, Failures: 8, Errors: 0, Skipped: 8
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

software-is-art commented:
@ahmetmircik / @vbekiaris I've updated the serialization binary file for the compatibility tests.

We've dropped the `Versioned` interface from `PartitionPredicateImpl` and I forgot that would cause issues with these tests.

vbekiaris commented:
run-lab-run

mafryer commented:
@ahmetmircik @vbekiaris is there anything more required for this PR?

mdumandag commented:
Looking at the recent discussion and changes, I believe the most important one is this one. We should be using `toDataWithStrategy` not `toData` here. https://github.com/hazelcast/hazelcast/pull/21319#discussion_r937410977

Also, we have this one, https://github.com/hazelcast/hazelcast/pull/21319#discussion_r939949318, but as Ahmet said, it can be done later in a separate PR

software-is-art commented:
> Looking at the recent discussion and changes, I believe the most important one is this one. We should be using `toDataWithStrategy` not `toData` here. [#21319 (comment)](https://github.com/hazelcast/hazelcast/pull/21319#discussion_r937410977)
> 
> Also, we have this one, [#21319 (comment)](https://github.com/hazelcast/hazelcast/pull/21319#discussion_r939949318), but as Ahmet said, it can be done later in a separate PR

I added a support for using `toDataWithStrategy` on a partition id set

ahmetmircik commented:
run-lab-run

ahmetmircik commented:
@mafryer please see this: https://github.com/hazelcast/hazelcast/pull/21319#discussion_r946488379

ahmetmircik reviewed (approved):
LGTM, thanks for the PR and addressing reviews patiently.

ahmetmircik commented:
run-lab-run

ahmetmircik commented:
@mdumandag are you ok with latest state of this PR?

mdumandag commented on hazelcast/src/main/java/com/hazelcast/query/Predicates.java:
```diff
@@ -571,10 +574,25 @@ public static <K, V> PagingPredicate<K, V> pagingPredicate(Predicate<K, V> predi
+     * @param <V>           the type of values the predicate operates on.
+     * @throws NullPointerException     if partitionKeys or target predicate are {@code null}
+     * @throws IllegalArgumentException if partitionkeys is an empty set
+     */
```
Minor: We can add `@since` tag here

mdumandag commented:
@ahmetmircik The latest state of the PR looks good to me. I just need to confirm one thing before approving the PR. Will we include this in 5.2? If not, we need to update the protocol PR with the correct versioning, but if we are going to include it, it can stay as it is, and I can approve the PR

ahmetmircik commented:
run-ee-tests

hz-devops-test commented:
The job `Hazelcast-pr-builder-ee-tests` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder-ee-tests/54/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder-ee-tests/54/Hazelcast-pr-builder-ee-tests-54.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=2048m; support was removed in 8.0
--------------------------
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   BPlusTreeMemoryTest.testDisposeWhileConcurrentlyAccessing:304 java.lang.AssertionError
[ERROR]   DeviceOperationExecutorTest.testExecutorShutdown:113->assertThreadCount:190 expected:<5> but was:<41>
[INFO] 
[ERROR] Tests run: 10454, Failures: 2, Errors: 0, Skipped: 86
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

software-is-art commented:
> LGTM, thanks for the PR and addressing reviews patiently.

Thanks @ahmetmircik for being patient with me as I worked through it all. Looking forward to upgrading to 5.2 when it's out.

All the best for the next release 👏
